### PR TITLE
Ride serialization v2

### PR DESCRIPTION
**Pull request by alexeykiselev**

### CODE REVIEW

File: pkg/ride/parserV2.go [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+			arguments[i] = arg
+		}
+		return NewFunctionCallNode(function, arguments), nil
+
+	case tokenBlockV2:
+		p.seenBlockV2 = true
+		declaration, err := p.readDeclaration()
+		if err != nil {
+			return nil, err
+		}
```
Comment (nickeskov) [inline pkg/ride/parserV2.go:235]: In `parserV1.go` I see this:
```go
declaration, err := p.readDeclaration()
if err != nil {
	return nil, err
}
if ad, ok := declaration.(*AssignmentNode); ok {
	ad.newBlock = true
}
```
But here I can't see this lines. Did you forget to add them?

**Review by nickeskov — CHANGES_REQUESTED** [review]

File: pkg/ride/meta/meta.go [hunk 2]
```diff
@@ -98,24 +105,97 @@ func Convert(meta *g.DAppMeta) (DApp, error) {
...
+		r.CompactNameAndOriginalNamePairList, r.OriginalNames = buildAbbreviations(m.Abbreviations)
+		return r, nil
 	default:
-		return nil, errors.Errorf("unsupported version of meta %d", version)
+		return nil, errors.Errorf("unsupported meta version %d", m.Version)
+	}
+}
+
+func convertAbbreviations(pairs []*g.DAppMeta_CompactNameAndOriginalNamePair, names []string) Abbreviations {
+	r := Abbreviations{pairs: make([]pair, len(pairs)), names: names, compact2original: make(map[string]string)}
```
[the remark this answers is not in the document]
Comment (alexeykiselev) [reply pkg/ride/meta/meta.go:190]: Will do

File: pkg/ride/serializerV1.go [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+		return errors.Errorf("unsupported type of node '%T'", node)
+	}
+}
+
+func (s *serializerV1) writeByte(b byte) error {
+	return s.buf.WriteByte(b)
+}
+
+func (s *serializerV1) writeUint16(v uint16) error {
+	b := make([]byte, 2)
```
[the remark this answers is not in the document]
Comment (alexeykiselev) [reply pkg/ride/serializerV1.go:343]: Yes, you right.
```golang
package array_slice_bench

import (
	"encoding/binary"
	"math"
	"testing"
)

var a [2]byte
var s []byte

func BenchmarkSlice(b *testing.B) {
	var v uint16 = math.MaxUint16
	for i := 0; i < b.N; i++ {
		s = make([]byte, 2)
		binary.BigEndian.PutUint16(s, v)
	}
}

func BenchmarkArray(b *testing.B) {
	var v uint16 = math.MaxUint16
	for i := 0; i < b.N; i++ {
		a = [2]byte{}
		binary.BigEndian.PutUint16(a[:], v)
	}
}
```
The array approach is way faster:
```
goos: darwin
goarch: amd64
pkg: array-slice-bench
cpu: Intel(R) Core(TM) i7-7820HQ CPU @ 2.90GHz
BenchmarkSlice
BenchmarkSlice-8   	95021604	        11.76 ns/op
BenchmarkArray
BenchmarkArray-8   	1000000000	         0.5911 ns/op
PASS
```
We have to use it everywhere.

File: pkg/ride/parserV2.go (hunk 1 above)
Comment (alexeykiselev) [reply pkg/ride/parserV2.go:235]: Yes, lost, will fix

**Review by nickeskov — DISMISSED** [review]

**Review by nickeskov — CHANGES_REQUESTED** [review]

**Review by nickeskov — APPROVED** [review]

**Review by nickeskov — APPROVED** [review]

**Pull request merged**